worlds/tloz/Client.py

```diff
@@ -0,0 +1,377 @@
+            write_list: list[tuple[int, list[int], str]] = [new_items_received_count_low_write, new_items_received_count_high_write]
+            await self.write_item(ctx, current_item_name, write_list)
+
+    async def resolve_shop_items(self, ctx: "BizHawkClientContext"):
+        left_shop_slots = await self.read_ram_value_guarded(ctx, Rom.left_shop_slots)
+        middle_shop_slots = await self.read_ram_value_guarded(ctx, Rom.middle_shop_slots)
+        right_shop_slots = await self.read_ram_value_guarded(ctx, Rom.right_shop_slots)
+        if left_shop_slots is None or middle_shop_slots is None or right_shop_slots is None:
+            return
+        for shop_category, shops in Locations.shop_categories.items():
+            bit = Rom.shop_correspondance[shop_category]
+            for shop in shops:
+                shop_id = Locations.location_table[shop] + base_id
+                if shop_id in ctx.checked_locations:
+                    if "Left" in shop:
+                        left_shop_slots = left_shop_slots | bit
+                    if "Middle" in shop:
+                        middle_shop_slots = middle_shop_slots | bit
+                    if "Right" in shop:
+                        right_shop_slots = right_shop_slots | bit
+        left_shop_write = (Rom.left_shop_slots, [left_shop_slots], self.wram)
+        middle_shop_write = (Rom.middle_shop_slots, [middle_shop_slots], self.wram)
+        right_shop_write = (Rom.right_shop_slots, [right_shop_slots], self.wram)
+        await self.write(ctx, [left_shop_write, middle_shop_write, right_shop_write])
+
+    async def resolve_triforce_fragments(self, ctx: "BizHawkClientContext"):
+        current_triforce_count = await self.read_ram_value_guarded(ctx, Rom.triforce_count)
+        if current_triforce_count is None:
+            return
+        current_triforce_byte = 0xFF >> (8 - min(current_triforce_count, 8))
+        write_list = [(Rom.triforce_fragments, [current_triforce_byte], self.wram)]
+        await self.write(ctx, write_list)
+
+    async def write_item(self, ctx, item_name, write_list: list[tuple[int, list[int], str]]):
+        item_game_id = Items.item_game_ids[item_name]
+        if item_name == "Bomb":
+            item_game_id = 0x29 # Hack to allow bombs to be shown being lifted.
+        lift_write = (Rom.item_to_lift, [item_game_id], self.wram)
+        timer_write = (Rom.item_lift_timer, [128], self.wram) # 128 frames of lifting an item
+        sound_write = (Rom.sound_effect_queue, [4], self.wram) # "Found secret" sound
+        write_list.extend([lift_write, timer_write, sound_write])
+        await self.handle_item(ctx, item_name, write_list)
+
+    async def resolve_bonus_items(self, ctx: "BizHawkClientContext"):
+        for item in self.bonus_items:
+            current_item_name = ctx.item_names.lookup_in_game(item, ctx.game)
+            await self.write_item(ctx, current_item_name, [])
+        self.bonus_items.clear()
```

Comment:
Last one for this batch.

If a guard fails while doing these writes (and it probably will if there's more than one item because of the same problem mentioned two comments up), then the bonus item it tried to give the player is lost. I think it's fixed if you only remove the item from this list if the guard doesn't fail, but I'll let you figure out the exact solution.

Reply:
Fixed in the latest commits, the item is only removed from the bonus list if the write succeeds, and that seemed to work fine in testing.